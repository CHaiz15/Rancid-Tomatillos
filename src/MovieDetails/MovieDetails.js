import React from 'react';
import './MovieDetails.scss';
import { connect } from 'react-redux';

const MovieDetails = ({ id, title, backdrop_path, release_date, overview, average_rating, user, ratings }) => {
  let foundRating = 0;
  
  if(!user.loggedOut && ratings.find(currentRating => currentRating.movie_id === id)) {
    foundRating = ratings.find(currentRating => currentRating.movie_id === id).rating
  }
  
  const ratingsDisplay = () => {
    if(user.loggedOut)  {
      return <h3>Avg. Rating: {average_rating}</h3>
    } else {
      return <div>
                <h3>Avg. Rating: {average_rating}</h3>
                  {foundRating === 0 ? 
                    <>
                      <h3>Add Rating: </h3>  
                        <div>
                          <select>
                            <option value="0">select rating:</option>
                            <option value="1">1</option>
                            <option value="2">2</option>
                            <option value="3">3</option>
                            <option value="4">4</option>
                            <option value="5">5</option>
                            <option value="6">6</option>
                            <option value="7">7</option>
                            <option value="8">8</option>
                            <option value="9">9</option>
                            <option value="10">10</option>
                          </select>
                        </div>
                      </>:
                    <h3>My Rating: {foundRating}</h3>
                  }
              </div>
    }
  }

  return (
    <article className='movie-details-pg'>
      <h2>{title}</h2>
      <p>Released: {release_date}</p>
      <h4>{ratingsDisplay()}</h4>
      <h4>{overview}</h4>
      <img className='detail-img' alt={title + 'image'}src={backdrop_path}/>
    </article>
  )
}

const mapStateToProps = state => ({
  ratings: state.ratings,
  user: state.user,
})

export default connect(mapStateToProps)(MovieDetails);