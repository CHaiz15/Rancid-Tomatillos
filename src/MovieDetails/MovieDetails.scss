.movie-details-pg {
  margin-top: 5rem;
  display: flex;
  @media (max-width: 750px) {
  }
  @media (max-width: 750px) {
    flex-direction: column;
    align-items: center;
  }
}

.float-img {
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  height: 50rem;
  width: 100%;
  position: absolute;
  z-index: -1;
  opacity: .08;
}

.movie-info {
  width: 50%;
  margin: 5%;
  min-width: 250px;
  @media (max-width: 1000px) {
    width: 80%;
  }
}

.detail-title {
  border-bottom: 1px solid black;
  margin-bottom: 2%;
}

.detail-release {
  margin-bottom: 5%;
}

.detail-overview {
  margin-bottom: 5%;
}

.detail-img {
  width: 30%;
  min-height: 350px;
  margin: 5%;
  box-shadow: 0px 6px 16px rgba(38, 38, 38, 0.6);
  @media (max-width: 750px) {
    width: 90%;
  }
}

.add-rating-box {
  display: flex;
  align-items: center;
}