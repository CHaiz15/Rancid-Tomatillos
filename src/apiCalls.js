export const fetchUser = async (email, password) => {

  const userInfo = {email, password};

  const options = {
    method: 'POST',
    body: JSON.stringify(userInfo),
    headers: {'Content-Type': 'application/json'}
  }

  const response = await fetch('https://rancid-tomatillos.herokuapp.com/api/v1/login', options);
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.message);
  }

  const user = await response.json();
  return user.user;
}

export const getMovies = () => {
  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/movies')
  .then(res => {
    if(!res.ok) {
      throw Error('Something is not right, try again later')
    }
    return res.json()})
}

export const getRatings = (id) => {
  return fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${id}/ratings`)
    .then(res => {
      if(!res.ok) {
        throw Error('Something is not right, try again later')
      }
      return res.json()})
}

export const postRating = async (rating, id, movie_id) => {
  const options = {
    method: 'POST',
    body: JSON.stringify({
      movie_id,
      rating,
    }),
    headers: {
      'Content-Type': 'application/json'  
    }
  }

  return await fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${id}/ratings`, options)
    .then(res => {
      if(!res.ok) {
        throw Error('Something is not right, try again later')
      }
      return res.json()})
}

export const deleteRating = async (rating_id, id) => {
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }

  return await fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${id}/ratings/${rating_id}`, options)
    .then(res => {
      if(!res.ok) {
        throw Error('Something is not right, try again later')
      }
    })
}