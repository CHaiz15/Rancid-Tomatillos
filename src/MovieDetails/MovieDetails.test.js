import React from 'react';
import { MovieDetails, mapStateToProps, mapDispatchToProps } from './MovieDetails';
import { shallow } from 'enzyme';
import { deleteRating } from '../apiCalls'

describe('MovieDetails', () => {
  let wrapper,mockUser,mockRatings,mockEvent,mockId
  beforeEach(() => {
    mockUser = {
      name: 'Ken',
      id: 23
    }
    mockRatings = [
      { movie_id: 23, name: 'Sonic', rating: 6 },
      { movie_id: 24, name: 'Parasite', rating: 9 }
    ]
    mockId = 23
    wrapper = shallow(<MovieDetails id={mockId} user={mockUser} ratings={mockRatings}/>);
  })
  it('should match the snapshot', () => {

    expect(wrapper).toMatchSnapshot();
  });

  it('should handleChange', () =>{
    mockEvent = {
      preventDefault: jest.fn(),
      target: {
        value: 7
      }
    }
    wrapper.instance().handleChange(mockEvent)
    expect(wrapper.state('currentMovieRating')).toEqual(7)
  })

  it('should call updateRatings on click', () =>{
    wrapper.instance().updateRatings = jest.fn()
    wrapper.find('input').simulate('click')
    expect(wrapper.instance().updateRatings).toHaveBeenCalled()
  })
});