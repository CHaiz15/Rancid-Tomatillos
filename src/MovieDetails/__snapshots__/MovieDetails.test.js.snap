// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails should match the snapshot 1`] = `
<article
  className="movie-details-pg"
>
  <div
    className="float-img"
    style={
      Object {
        "backgroundImage": "url(undefined)",
      }
    }
  />
  <section
    className="movie-info"
  >
    <h2
      className="detail-title"
    />
    <p
      className="detail-release"
    >
      Released: 
    </p>
    <h4
      className="detail-overview"
    />
    <div
      className="ratings-container"
    >
      <h3
        className="rating"
      >
        Avg. Rating:  
        NaN
      </h3>
      <div>
        <h3
          className="rating"
        >
          My Rating:  
          6
        </h3>
        <form
          className="add-rating"
        >
          <h3
            className="rating"
          >
            Add Rating: 
          </h3>
          <div
            className="add-rating-box"
          >
            <select
              className="rating-select"
              onChange={[Function]}
              value={6}
            >
              <option
                value="0"
              >
                -
              </option>
              <option
                value="1"
              >
                1
              </option>
              <option
                value="2"
              >
                2
              </option>
              <option
                value="3"
              >
                3
              </option>
              <option
                value="4"
              >
                4
              </option>
              <option
                value="5"
              >
                5
              </option>
              <option
                value="6"
              >
                6
              </option>
              <option
                value="7"
              >
                7
              </option>
              <option
                value="8"
              >
                8
              </option>
              <option
                value="9"
              >
                9
              </option>
              <option
                value="10"
              >
                10
              </option>
            </select>
            <input
              className="rate-btn"
              onClick={[Function]}
              type="button"
              value="Rate"
            />
          </div>
        </form>
      </div>
    </div>
  </section>
  <img
    alt="undefinedimage"
    className="detail-img"
  />
</article>
`;
